docs/Execution.md

```diff
@@ -0,0 +1,107 @@
+
+- It simplifies MsQuic's logic by eliminating the need for a queue or cached state to indicate to the application. In the BSD model, the networking stack must wait for a top-down call from the application before indicating completion, increasing code size, complexity, and memory usage.
+
+### Writing Event Handlers
+
+Event handlers are **essential** for all objects that support them, as much of the MsQuic API operates through these callbacks.
+Critical events, such as "shutdown complete" notifications, provide vital information necessary for the application to function correctly.
+Without these events, the application cannot determine when it is safe to clean up objects.
+
+Applications should keep the execution time within callbacks **to a minimum**.
+MsQuic does not use separate threads for protocol execution and upcalls to the application.
+Therefore, any significant delays in the callback **will delay the protocol**.
+Any substantial work required by the application should be performed on its own thread.
+
+This does not mean the application cannot perform any work in the callback handler.
+In fact, many operations are designed to be most efficient when executed within the callback.
+For example, closing a handle to a connection or stream is ideally done during the "shutdown complete" indication.
+
+A crucial aspect of this design is that all blocking API (down) calls invoked within a callback always occur inline (to prevent deadlocks) and will take precedence over any calls in progress or queued from a separate thread.
+
+## Threading
+
+By default, MsQuic creates its own threads to manage the execution of its logic.
+The number and configuration of these threads depend on the settings passed to [RegistrationOpen](api/RegistrationOpen.md) or `QUIC_PARAM_GLOBAL_EXECUTION_CONFIG`.
+
+Typically, MsQuic creates dedicated threads for each processor, which are hard-affinitized to a specific NUMA node and soft-affinitized (set as 'ideal processor') to a specific processor.
+These threads handle both the datapath (i.e., UDP) and QUIC layers.
+MsQuic aligns its processing logic with the rest of the networking stack (including hardware RSS) to ensure that all processing stays on the same NUMA node, and ideally, the same processor.
+
+The complexity of aligning processing across various threads and processors is why MsQuic manages its own threading by default.
+This abstraction simplifies the development process for applications built on top of MsQuic, ensuring that things "just work" out of the box.
+
+Each thread manages the execution of one or more connections.
+Connections are distributed across threads based on their RSS alignment, which should evenly distribute traffic based on different UDP tuples.
```

Comment:
except RSS hashing isn't always evenly distributed :(
so maybe it's more correct to remove "which should be evenly distributed"

Reply:
Well, we already say "which **should** evenly distribute traffic". 😄 In the future, we plan to add QUIC-based RSS which will allow us explicit control over this, so we will be able to guarantee it.